File: src/main/java/org/trp/shincolle/menu/CraneMenu.java
package org.trp.shincolle.menu;

import net.minecraft.core.BlockPos;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.ClickType;
import net.minecraft.world.inventory.DataSlot;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.neoforged.neoforge.items.SlotItemHandler;
import org.trp.shincolle.block.entity.CraneBlockEntity;
import org.trp.shincolle.entity.base.EntityShipBase;

public class CraneMenu extends AbstractContainerMenu {
    private final CraneBlockEntity blockEntity;
    private final boolean clientSide;
    private final Level level;

    
    private int remainedPower;
    private int powerMax;
    private int packedBooleans; 
    private int craneMode;
    private int modeItem;
    private int modeRedstone;
    private int modeLiquid;
    private int modeEnergy;
    private int shipEntityId;
    private int shipTimer;

    public CraneMenu(int containerId, Inventory playerInventory, RegistryFriendlyByteBuf buffer) {
        this(containerId, playerInventory, getBlockEntity(playerInventory, buffer));
    }

    public CraneMenu(int containerId, Inventory playerInventory, CraneBlockEntity blockEntity) {
        super(ModMenus.CRANE_MENU.get(), containerId);
        this.blockEntity = blockEntity;
        this.clientSide = playerInventory.player.level().isClientSide;
        this.level = playerInventory.player.level();

        if (!this.clientSide && blockEntity.getOwnerUUID() == null) {
            blockEntity.setOwnerUUID(playerInventory.player.getUUID());
            blockEntity.setOwnerName(playerInventory.player.getName().getString());
        }

        
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getRemainedPower(); }
            @Override public void set(int value) { remainedPower = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getPowerMax(); }
            @Override public void set(int value) { powerMax = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() {
                int packed = 0;
                if (blockEntity.isActive()) packed |= 1;
                if (blockEntity.isCheckMetadata()) packed |= 2;
                if (blockEntity.isCheckOredict()) packed |= 4;
                if (blockEntity.isCheckNbt()) packed |= 8;
                if (blockEntity.isEnabLoad()) packed |= 16;
                if (blockEntity.isEnabUnload()) packed |= 32;
                return packed;
            }
            @Override public void set(int value) { packedBooleans = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getCraneMode(); }
            @Override public void set(int value) { craneMode = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getModeItem(); }
            @Override public void set(int value) { modeItem = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getModeRedstone(); }
            @Override public void set(int value) { modeRedstone = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getModeLiquid(); }
            @Override public void set(int value) { modeLiquid = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getModeEnergy(); }
            @Override public void set(int value) { modeEnergy = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getCraningShipId(); }
            @Override public void set(int value) { shipEntityId = value; }
        });
        this.addDataSlot(new DataSlot() {
            @Override public int get() { return blockEntity.getCraningShipTimer(); }
            @Override public void set(int value) { shipTimer = value; }
        });

        
        for (int i = 0; i < 9; ++i) {
            this.addSlot(new SlotItemHandler(blockEntity.getInventory(), i, 8 + i * 18, 65));
        }
        for (int i = 0; i < 9; ++i) {
            this.addSlot(new SlotItemHandler(blockEntity.getInventory(), i + 9, 8 + i * 18, 96));
        }

        
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 9; ++j) {
                this.addSlot(new Slot(playerInventory, j + i * 9 + 9, 8 + j * 18, 119 + i * 18));
            }
        }
        for (int i = 0; i < 9; ++i) {
            this.addSlot(new Slot(playerInventory, i, 8 + i * 18, 177));
        }
    }

    @Override
    public boolean stillValid(Player player) {
        return player.distanceToSqr(blockEntity.getBlockPos().getX() + 0.5, blockEntity.getBlockPos().getY() + 0.5, blockEntity.getBlockPos().getZ() + 0.5) <= 64.0;
    }

    @Override
    public ItemStack quickMoveStack(Player player, int index) {
        
        
        return ItemStack.EMPTY;
    }

    @Override
    public void clicked(int slotId, int button, ClickType clickType, Player player) {
        
        if (slotId >= 0 && slotId < 18) {
            ItemStack held = player.containerMenu.getCarried();
            Slot slot = this.slots.get(slotId);
            
            if (!held.isEmpty()) {
                ItemStack ghost = held.copy();
                if (button == 1) { 
                    ghost.setCount(1);
                    slot.set(ghost);
                    blockEntity.setItemMode(slotId, true);
                } else { 
                    
                    ItemStack old = slot.getItem();
                    if (ItemStack.isSameItem(ghost, old)) {
                        ghost.setCount(Math.min(ghost.getMaxStackSize(), ghost.getCount() + old.getCount()));
                    }
                    slot.set(ghost);
                    blockEntity.setItemMode(slotId, false);
                }
            } else {
                slot.set(ItemStack.EMPTY);
                blockEntity.setItemMode(slotId, false);
            }
            return;
        }
        super.clicked(slotId, button, clickType, player);
    }

    @Override
    public boolean clickMenuButton(Player player, int id) {
        switch (id) {
            case 0: 
                blockEntity.setActive(!blockEntity.isActive());
                return true;
            case 1: 
                int m = blockEntity.getCraneMode() + 1;
                if (m > 24) m = 24;
                blockEntity.setCraneMode(m);
                return true;
            case 11: 
                int m2 = blockEntity.getCraneMode() - 1;
                if (m2 < 0) m2 = 0;
                blockEntity.setCraneMode(m2);
                return true;
            case 2: 
                blockEntity.setCheckMetadata(!blockEntity.isCheckMetadata());
                return true;
            case 3: 
                blockEntity.setCheckOredict(!blockEntity.isCheckOredict());
                return true;
            case 6: 
                blockEntity.setCheckNbt(!blockEntity.isCheckNbt());
                return true;
            case 7: 
                int r = (blockEntity.getModeRedstone() + 1) % 3;
                blockEntity.setModeRedstone(r);
                return true;
            case 4: 
                blockEntity.setEnabLoad(!blockEntity.isEnabLoad());
                return true;
            case 5: 
                blockEntity.setEnabUnload(!blockEntity.isEnabUnload());
                return true;
            case 8: 
                int l = (blockEntity.getModeLiquid() + 1) % 3;
                blockEntity.setModeLiquid(l);
                return true;
        }
        return super.clickMenuButton(player, id);
    }

    
    public int getRemainedPower() { return clientSide ? remainedPower : blockEntity.getRemainedPower(); }
    public int getPowerMax() { return clientSide ? powerMax : blockEntity.getPowerMax(); }
    public boolean isActive() { 
        int val = clientSide ? packedBooleans : (
            (blockEntity.isActive() ? 1 : 0) |
            (blockEntity.isCheckMetadata() ? 2 : 0) |
            (blockEntity.isCheckOredict() ? 4 : 0) |
            (blockEntity.isCheckNbt() ? 8 : 0) |
            (blockEntity.isEnabLoad() ? 16 : 0) |
            (blockEntity.isEnabUnload() ? 32 : 0)
        );
        return (val & 1) != 0; 
    }
    public boolean isCheckMetadata() { 
        int val = clientSide ? packedBooleans : (
            (blockEntity.isActive() ? 1 : 0) |
            (blockEntity.isCheckMetadata() ? 2 : 0) |
            (blockEntity.isCheckOredict() ? 4 : 0) |
            (blockEntity.isCheckNbt() ? 8 : 0) |
            (blockEntity.isEnabLoad() ? 16 : 0) |
            (blockEntity.isEnabUnload() ? 32 : 0)
        );
        return (val & 2) != 0; 
    }
    public boolean isCheckOredict() { 
        int val = clientSide ? packedBooleans : (
            (blockEntity.isActive() ? 1 : 0) |
            (blockEntity.isCheckMetadata() ? 2 : 0) |
            (blockEntity.isCheckOredict() ? 4 : 0) |
            (blockEntity.isCheckNbt() ? 8 : 0) |
            (blockEntity.isEnabLoad() ? 16 : 0) |
            (blockEntity.isEnabUnload() ? 32 : 0)
        );
        return (val & 4) != 0; 
    }
    public boolean isCheckNbt() { 
        int val = clientSide ? packedBooleans : (
            (blockEntity.isActive() ? 1 : 0) |
            (blockEntity.isCheckMetadata() ? 2 : 0) |
            (blockEntity.isCheckOredict() ? 4 : 0) |
            (blockEntity.isCheckNbt() ? 8 : 0) |
            (blockEntity.isEnabLoad() ? 16 : 0) |
            (blockEntity.isEnabUnload() ? 32 : 0)
        );
        return (val & 8) != 0; 
    }
    public boolean isEnabLoad() { 
        int val = clientSide ? packedBooleans : (
            (blockEntity.isActive() ? 1 : 0) |
            (blockEntity.isCheckMetadata() ? 2 : 0) |
            (blockEntity.isCheckOredict() ? 4 : 0) |
            (blockEntity.isCheckNbt() ? 8 : 0) |
            (blockEntity.isEnabLoad() ? 16 : 0) |
            (blockEntity.isEnabUnload() ? 32 : 0)
        );
        return (val & 16) != 0; 
    }
    public boolean isEnabUnload() { 
        int val = clientSide ? packedBooleans : (
            (blockEntity.isActive() ? 1 : 0) |
            (blockEntity.isCheckMetadata() ? 2 : 0) |
            (blockEntity.isCheckOredict() ? 4 : 0) |
            (blockEntity.isCheckNbt() ? 8 : 0) |
            (blockEntity.isEnabLoad() ? 16 : 0) |
            (blockEntity.isEnabUnload() ? 32 : 0)
        );
        return (val & 32) != 0; 
    }
    public int getCraneMode() { return clientSide ? craneMode : blockEntity.getCraneMode(); }
    public int getModeItem() { return clientSide ? modeItem : blockEntity.getModeItem(); }
    public int getModeRedstone() { return clientSide ? modeRedstone : blockEntity.getModeRedstone(); }
    public int getModeLiquid() { return clientSide ? modeLiquid : blockEntity.getModeLiquid(); }
    public int getModeEnergy() { return clientSide ? modeEnergy : blockEntity.getModeEnergy(); }

    public int getShipEntityId() { return clientSide ? shipEntityId : blockEntity.getCraningShipId(); }
    public int getShipTimer() { return clientSide ? shipTimer : blockEntity.getCraningShipTimer(); }

    public EntityShipBase getShipEntity() {
        int id = getShipEntityId();
        if (id <= 0) {
            return null;
        }
        if (level.getEntity(id) instanceof EntityShipBase ship) {
            return ship;
        }
        return null;
    }

    public CraneBlockEntity getBlockEntity() { return blockEntity; }

    private static CraneBlockEntity getBlockEntity(Inventory playerInventory, RegistryFriendlyByteBuf buffer) {
        BlockPos pos = buffer.readBlockPos();
        if (playerInventory.player.level().getBlockEntity(pos) instanceof CraneBlockEntity crane) {
            return crane;
        }
        throw new IllegalStateException("Crane block entity not found.");
    }
}

File: src/main/java/org/trp/shincolle/menu/CraneMenu.kt
package org.trp.shincolle.menu

import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ClickType
import net.minecraft.world.inventory.DataSlot
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.neoforged.neoforge.items.SlotItemHandler
import org.trp.shincolle.block.entity.CraneBlockEntity
import org.trp.shincolle.entity.base.EntityShipBase
import kotlin.math.min

class CraneMenu(containerId: Int, playerInventory: Inventory, val blockEntity: CraneBlockEntity) :
    AbstractContainerMenu(ModMenus.CRANE_MENU.get(), containerId) {
    private val clientSide: Boolean
    private val level: Level


    private var remainedPower = 0
    private var powerMax = 0
    private var packedBooleans = 0
    private var craneMode = 0
    private var modeItem = 0
    private var modeRedstone = 0
    private var modeLiquid = 0
    private var modeEnergy = 0
    private var shipEntityId = 0
    private var shipTimer = 0

    constructor(containerId: Int, playerInventory: Inventory, buffer: RegistryFriendlyByteBuf) : this(
        containerId,
        playerInventory,
        getBlockEntity(playerInventory, buffer)
    )

    init {
        this.clientSide = playerInventory.player.level().isClientSide
        this.level = playerInventory.player.level()

        if (!this.clientSide && blockEntity.getOwnerUUID() == null) {
            blockEntity.setOwnerUUID(playerInventory.player.getUUID())
            blockEntity.setOwnerName(playerInventory.player.getName().getString())
        }


        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getRemainedPower()
            }

            override fun set(value: Int) {
                remainedPower = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getPowerMax()
            }

            override fun set(value: Int) {
                powerMax = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                var packed = 0
                if (blockEntity.isActive()) packed = packed or 1
                if (blockEntity.isCheckMetadata()) packed = packed or 2
                if (blockEntity.isCheckOredict()) packed = packed or 4
                if (blockEntity.isCheckNbt()) packed = packed or 8
                if (blockEntity.isEnabLoad()) packed = packed or 16
                if (blockEntity.isEnabUnload()) packed = packed or 32
                return packed
            }

            override fun set(value: Int) {
                packedBooleans = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getCraneMode()
            }

            override fun set(value: Int) {
                craneMode = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getModeItem()
            }

            override fun set(value: Int) {
                modeItem = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getModeRedstone()
            }

            override fun set(value: Int) {
                modeRedstone = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getModeLiquid()
            }

            override fun set(value: Int) {
                modeLiquid = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getModeEnergy()
            }

            override fun set(value: Int) {
                modeEnergy = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getCraningShipId()
            }

            override fun set(value: Int) {
                shipEntityId = value
            }
        })
        this.addDataSlot(object : DataSlot() {
            override fun get(): Int {
                return blockEntity.getCraningShipTimer()
            }

            override fun set(value: Int) {
                shipTimer = value
            }
        })


        for (i in 0..8) {
            this.addSlot(SlotItemHandler(blockEntity.inventory, i, 8 + i * 18, 65))
        }
        for (i in 0..8) {
            this.addSlot(SlotItemHandler(blockEntity.inventory, i + 9, 8 + i * 18, 96))
        }


        for (i in 0..2) {
            for (j in 0..8) {
                this.addSlot(Slot(playerInventory, j + i * 9 + 9, 8 + j * 18, 119 + i * 18))
            }
        }
        for (i in 0..8) {
            this.addSlot(Slot(playerInventory, i, 8 + i * 18, 177))
        }
    }

    override fun stillValid(player: Player): Boolean {
        if (this.blockEntity.getLevel() == null) {
            return false
        }
        if (player.level().getBlockEntity(this.blockEntity.getBlockPos()) !== this.blockEntity) {
            return false
        }
        return player.distanceToSqr(
            blockEntity.getBlockPos().getX() + 0.5,
            blockEntity.getBlockPos().getY() + 0.5,
            blockEntity.getBlockPos().getZ() + 0.5
        ) <= 64.0
    }

    override fun quickMoveStack(player: Player, index: Int): ItemStack {
        // Legacy crane UI intentionally disables shift-transfer because the first 18 slots
        // are ghost filters rather than real storage destinations.
        return ItemStack.EMPTY
    }

    override fun clicked(slotId: Int, button: Int, clickType: ClickType, player: Player) {
        if (player.level().isClientSide) {
            return
        }

        if (slotId >= 0 && slotId < 18) {
            val held = player.containerMenu.getCarried()
            val slot = this.slots.get(slotId)

            if (!held.isEmpty()) {
                val ghost = held.copy()
                if (button == 1) {
                    ghost.setCount(1)
                    slot.set(ghost)
                    blockEntity.setItemMode(slotId, true)
                } else {
                    val old = slot.getItem()
                    if (ItemStack.isSameItem(ghost, old)) {
                        ghost.setCount(min(ghost.getMaxStackSize(), ghost.getCount() + old.getCount()))
                    }
                    slot.set(ghost)
                    blockEntity.setItemMode(slotId, false)
                }
            } else {
                slot.set(ItemStack.EMPTY)
                blockEntity.setItemMode(slotId, false)
            }
            return
        }
        super.clicked(slotId, button, clickType, player)
    }

    override fun clickMenuButton(player: Player, id: Int): Boolean {
        if (player.level().isClientSide) {
            return true
        }

        when (id) {
            0 -> {
                blockEntity.setActive(!blockEntity.isActive())
                return true
            }

            1 -> {
                val m = blockEntity.getCraneMode() + 1
                if (m > 24) m = 24
                blockEntity.setCraneMode(m)
                return true
            }

            11 -> {
                val m2 = blockEntity.getCraneMode() - 1
                if (m2 < 0) m2 = 0
                blockEntity.setCraneMode(m2)
                return true
            }

            2 -> {
                blockEntity.setCheckMetadata(!blockEntity.isCheckMetadata())
                return true
            }

            3 -> {
                blockEntity.setCheckOredict(!blockEntity.isCheckOredict())
                return true
            }

            6 -> {
                blockEntity.setCheckNbt(!blockEntity.isCheckNbt())
                return true
            }

            7 -> {
                val r = (blockEntity.getModeRedstone() + 1) % 3
                blockEntity.setModeRedstone(r)
                return true
            }

            4 -> {
                blockEntity.setEnabLoad(!blockEntity.isEnabLoad())
                return true
            }

            5 -> {
                blockEntity.setEnabUnload(!blockEntity.isEnabUnload())
                return true
            }

            8 -> {
                val l = (blockEntity.getModeLiquid() + 1) % 3
                blockEntity.setModeLiquid(l)
                return true
            }

            9 -> {
                val e = (blockEntity.getModeEnergy() + 1) % 3
                blockEntity.setModeEnergy(e)
                return true
            }
        }
        return super.clickMenuButton(player, id)
    }


    fun getRemainedPower(): Int {
        return if (clientSide) remainedPower else blockEntity.getRemainedPower()
    }

    fun getPowerMax(): Int {
        return if (clientSide) powerMax else blockEntity.getPowerMax()
    }

    val isActive: Boolean
        get() {
            val `val` = if (clientSide) packedBooleans else ((if (blockEntity.isActive()) 1 else 0) or
                    (if (blockEntity.isCheckMetadata()) 2 else 0) or
                    (if (blockEntity.isCheckOredict()) 4 else 0) or
                    (if (blockEntity.isCheckNbt()) 8 else 0) or
                    (if (blockEntity.isEnabLoad()) 16 else 0) or
                    (if (blockEntity.isEnabUnload()) 32 else 0)
                    )
            return (`val` and 1) != 0
        }
    val isCheckMetadata: Boolean
        get() {
            val `val` = if (clientSide) packedBooleans else ((if (blockEntity.isActive()) 1 else 0) or
                    (if (blockEntity.isCheckMetadata()) 2 else 0) or
                    (if (blockEntity.isCheckOredict()) 4 else 0) or
                    (if (blockEntity.isCheckNbt()) 8 else 0) or
                    (if (blockEntity.isEnabLoad()) 16 else 0) or
                    (if (blockEntity.isEnabUnload()) 32 else 0)
                    )
            return (`val` and 2) != 0
        }
    val isCheckOredict: Boolean
        get() {
            val `val` = if (clientSide) packedBooleans else ((if (blockEntity.isActive()) 1 else 0) or
                    (if (blockEntity.isCheckMetadata()) 2 else 0) or
                    (if (blockEntity.isCheckOredict()) 4 else 0) or
                    (if (blockEntity.isCheckNbt()) 8 else 0) or
                    (if (blockEntity.isEnabLoad()) 16 else 0) or
                    (if (blockEntity.isEnabUnload()) 32 else 0)
                    )
            return (`val` and 4) != 0
        }
    val isCheckNbt: Boolean
        get() {
            val `val` = if (clientSide) packedBooleans else ((if (blockEntity.isActive()) 1 else 0) or
                    (if (blockEntity.isCheckMetadata()) 2 else 0) or
                    (if (blockEntity.isCheckOredict()) 4 else 0) or
                    (if (blockEntity.isCheckNbt()) 8 else 0) or
                    (if (blockEntity.isEnabLoad()) 16 else 0) or
                    (if (blockEntity.isEnabUnload()) 32 else 0)
                    )
            return (`val` and 8) != 0
        }
    val isEnabLoad: Boolean
        get() {
            val `val` = if (clientSide) packedBooleans else ((if (blockEntity.isActive()) 1 else 0) or
                    (if (blockEntity.isCheckMetadata()) 2 else 0) or
                    (if (blockEntity.isCheckOredict()) 4 else 0) or
                    (if (blockEntity.isCheckNbt()) 8 else 0) or
                    (if (blockEntity.isEnabLoad()) 16 else 0) or
                    (if (blockEntity.isEnabUnload()) 32 else 0)
                    )
            return (`val` and 16) != 0
        }
    val isEnabUnload: Boolean
        get() {
            val `val` = if (clientSide) packedBooleans else ((if (blockEntity.isActive()) 1 else 0) or
                    (if (blockEntity.isCheckMetadata()) 2 else 0) or
                    (if (blockEntity.isCheckOredict()) 4 else 0) or
                    (if (blockEntity.isCheckNbt()) 8 else 0) or
                    (if (blockEntity.isEnabLoad()) 16 else 0) or
                    (if (blockEntity.isEnabUnload()) 32 else 0)
                    )
            return (`val` and 32) != 0
        }

    fun getCraneMode(): Int {
        return if (clientSide) craneMode else blockEntity.getCraneMode()
    }

    fun getModeItem(): Int {
        return if (clientSide) modeItem else blockEntity.getModeItem()
    }

    fun getItemMode(slotId: Int): Boolean {
        return (getModeItem() and (1 shl slotId)) != 0
    }

    fun getModeRedstone(): Int {
        return if (clientSide) modeRedstone else blockEntity.getModeRedstone()
    }

    fun getModeLiquid(): Int {
        return if (clientSide) modeLiquid else blockEntity.getModeLiquid()
    }

    fun getModeEnergy(): Int {
        return if (clientSide) modeEnergy else blockEntity.getModeEnergy()
    }

    fun getShipEntityId(): Int {
        return if (clientSide) shipEntityId else blockEntity.getCraningShipId()
    }

    fun getShipTimer(): Int {
        return if (clientSide) shipTimer else blockEntity.getCraningShipTimer()
    }

    val shipEntity: EntityShipBase?
        get() {
            val id = getShipEntityId()
            if (id <= 0) {
                return null
            }
            if (level.getEntity(id) is EntityShipBase
                && ship.isAlive()
                && !ship.isRemoved()
            ) {
                return ship
            }
            return null
        }

    companion object {
        private fun getBlockEntity(playerInventory: Inventory, buffer: RegistryFriendlyByteBuf): CraneBlockEntity {
            checkNotNull(buffer) { "Missing crane menu data." }
            val pos = buffer.readBlockPos()
            if (playerInventory.player.level().getBlockEntity(pos) is CraneBlockEntity) {
                return crane
            }
            throw IllegalStateException("Crane block entity not found.")
        }
    }
}
